src/tracing/marching_squares.cpp

```diff
@@ -141,23 +152,121 @@ static void marching_squares(bool topLeft, bool topRight, bool bottomRight, bool
+ */
+static void trace_contour(const std::size_t vertex_index, std::vector<vertex>& vertices, const std::vector<line>& lines, std::vector<std::size_t>& out_contour) {
+    // Add vertex to contour.
+    out_contour.push_back(vertex_index);
+
+    // Find connected vertex on line.
+    const line& line = lines[vertices[vertex_index].line_index];
+    std::size_t connected_vertex_index = line.vertex_indices[0] == vertex_index ? line.vertex_indices[1] : line.vertex_indices[0];
+    
+    // Set vertices as assigned to contour.
+    vertices[vertex_index].assigned_contour = true;
+    vertices[connected_vertex_index].assigned_contour = true;
+
+    // Find other vertex starting at connected vertex.
+    std::size_t found_vertex_index;
+    find_equal_vertex(connected_vertex_index, vertices, found_vertex_index);
+
+    // Check whether found vertex is assigned.
+    if (!vertices[found_vertex_index].assigned_contour) {
+        // Continue to trace contour.
+        trace_contour(found_vertex_index, vertices, lines, out_contour);
+    } else {
+        // Finnish contour by connecting tail and root.
+        out_contour.push_back(out_contour.front());
+    }
+}
+
 void trace_marching_squares(bool* layer, unsigned int layer_width, unsigned int layer_height, ptg_outline*& out_outlines, unsigned int& out_outline_count) {
 
     std::vector<line> lines;
-    std::vector<ptg_vec2> vertices;
+    std::vector<vertex> vertices;
 
     // Execute marching squares on layer.
-    for (unsigned int it = 0; it < layer_width * (layer_height - 1); ++it)
-        if ((it + 1) % layer_width != 0)
-            marching_squares(layer[it], layer[it + 1], layer[it + layer_width], layer[it + layer_width + 1], it % layer_width + 1, it / layer_width + 1, lines, vertices);
+    for (unsigned int it = 0; it < (layer_width * layer_height); ++it) {
+        bool topEdge = it < layer_width;
+        bool bottomEdge = it >= (layer_width * (layer_height - 1));
+        bool leftEdge = it % layer_width == 0;
+        bool rightEdge = it % layer_width == (layer_width - 1);
+
+        bool topLeft = topEdge || leftEdge ? false : layer[it - layer_width - 1];
+        bool topRight = topEdge || rightEdge ? false : layer[it - layer_width];
+        bool bottomRight = bottomEdge || rightEdge ? false : layer[it];
+        bool bottomLeft = bottomEdge || leftEdge ? false : layer[it - 1];
+
+        marching_squares(topLeft, topRight, bottomRight, bottomLeft, it % layer_width, it / layer_width, lines, vertices);
+    }
+
+    // Create contours.
+    std::vector<std::vector<std::size_t>> contours;
+    for (std::size_t vertex_index = 0; vertex_index < vertices.size(); ++vertex_index) {
+        if (!vertices[vertex_index].assigned_contour) {
+
+            // Create new contour.
+            contours.push_back(std::vector<std::size_t>());
+            std::vector<std::size_t>& contour = contours.back();
+
+            // Trace contour (assign connected vertex indices to contour).
```

Comment:
Duplicate code that already exists in trace_contour?

Reply:
Duplicate code was removed